src/mgmt/srv_pool.c

```diff
@@ -616,37 +619,53 @@ enum_pool_cb(daos_handle_t ih, d_iov_t *key, d_iov_t *val, void *varg)
 		return -DER_IO;
 	rec = val->iov_buf;
 
-	size = 4096;
-	D_ALLOC(buf, size);
-	if (buf == NULL)
-		return -DER_NOMEM;
-	p = buf;
-	for (i = 0; i < rec->pr_nreplicas; i++) {
-		int n;
-
-		n = snprintf(p, buf + size - p, "%s%u", i == 0 ? "" : ",",
-			     rec->pr_replicas[i]);
-		if (n < 0 || n >= buf + size - p) {
-			D_FREE(buf);
-			return -DER_OVERFLOW;
-		}
-		p += n;
-	}
+	ap->npools++;
+
+	/* Client didn't provide buffer or not enough space */
+	if (ap->avail_npools < ap->npools)
+		return 0;
 
-	D_DEBUG(DB_MGMT, "  "DF_UUID": state=%u svc=%s\n",
-		DP_UUID(key->iov_buf), rec->pr_state, buf);
+	/* Realloc pools[] if needed (double each time starting with 1) */
+	if (ap->pools_index == ap->pools_len) {
+		void	*ptr;
+		size_t	realloc_bytes;
 
-	D_FREE(buf);
+		realloc_bytes = ((1 << ap->pools_len) *
```

Comment:
Did you mean ap->pools_len << 1? If so, I'd personally just use the stupid way: ap->pools_len * 2.

Reply:
I had a bug at first with some simpler logic. pools_len starts at 0. Given this I want the first (re)alloc to be of length 1, then proceed with doubling after that.

Reply:
Hmm, sorry, I'm being slow... According to the comment above, I'd expect the number of elements in the array to be:

     1  2  4  8 16 ...

But we shall actually get:

     1  2  4 16 64K ...

No?

Reply:
You are right this needs to be simplified and corrected.